### PR TITLE
V0.2.5

### PR DESCRIPTION
**Pull request by hakancelik96**
- 💪 Refactor code rewrite using libcst.
- 🐞 Refactor bugs fix.
- 🧪 [Comma rare case support & test.](https://github.com/hakancelik96/unimport/blob/master/rare_cases/case_comma.py)
- 💪 pre-commit add
- 🐞 Add Sytranx Error Catcher

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -29,18 +29,17 @@ def get_long_description():
     license="MIT",
     python_requires=">=3.6.0",
     packages=["unimport"],
-    extras_require={"pyproject": ["toml"]},
+    install_requires=["libcst==0.3.4"],
```
Comment (isidentical) [inline setup.py:32]: This should be `0.3.5`

File: setup.py [hunk 2]
```diff
@@ -29,18 +29,17 @@ def get_long_description():
...
+    extras_require={"dev": ["toml"]},
```
Comment (isidentical) [inline setup.py:33]: There is no need to change the name, it should remain as `pyproject`

File: unimport/refactor.py [hunk 3]
```diff
@@ -1,137 +1,113 @@
...
+from unimport.scan import Scanner
+
+
+class RemoveUnusedImportTransformer(cst.CSTTransformer):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+
+    def __init__(self, source):
+        self.scanner = Scanner(source)
+        self.unused_imports = list(self.scanner.get_unused_imports())
+        self.imports = self.scanner.imports
```
Comment (isidentical) [inline unimport/refactor.py:13]: This shouldn't instantiate the scanner, it can take relevant data from `Session`.

File: requirements-dev.txt [hunk 4]
```diff
@@ -1,2 +1,5 @@
 toml==0.10.0
 pytest
+pre-commit==2.4.0
+black==19.10b0
+isort==4.3.21
```
Comment (isidentical) [inline requirements-dev.txt:5]: these are not needed

File: setup.py [hunk 5]
```diff
@@ -29,18 +29,17 @@ def get_long_description():
...
     include_package_data=True,
     classifiers=[
         "License :: OSI Approved :: MIT License",
+        "Topic :: Software Development :: Libraries :: Python Modules",
         "Environment :: Console",
-        "Programming Language :: Python",
         "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
-        "Programming Language :: Python :: 3 :: Only",
```
Comment (isidentical) [inline setup.py:42]: why was this deleted

File: unimport/refactor.py [hunk 6]
```diff
@@ -1,137 +1,111 @@
...
+from libcst.metadata import MetadataWrapper, PositionProvider
+
+
+class RemoveUnusedImportTransformer(cst.CSTTransformer):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+
+    def __init__(self, scanner):
+        self.scanner = scanner
+        self.unused_imports = list(self.scanner.get_unused_imports())
+        self.imports = self.scanner.imports
```
Comment (isidentical) [inline unimport/refactor.py:11]: this shouldn't take `scanner`, it should take the `unused_imports` as a list from the session

File: unimport/refactor.py [hunk 7]
```diff
@@ -1,137 +1,111 @@
...
+        module_attr.append(children[2].value)  # last value
+        node_value = children[0]
+        while True:
+            if hasattr(node_value, "attr"):
+                module_attr.append(node_value.attr.value)
+                if hasattr(node_value, "value"):  # get parent
+                    node_value = node_value.value
+                else:
+                    module_attr.append(node_value.value.value)
+                    break
```
Comment (isidentical) [inline unimport/refactor.py:25]: wow, this looks soo complex. please simplify it, and if you can get rid of this

File: unimport/refactor.py [hunk 8]
```diff
@@ -1,137 +1,111 @@
...
-            children.append(trailing_comma)
+    def is_import_used(self, import_name, location):
+        for imp in self.unused_imports:
+            if (
+                imp["name"] == import_name
+                and imp["lineno"] == location.start.line
+            ):
+                # unused
+                return False
+        return True
```
Comment (isidentical) [inline unimport/refactor.py:40]: I feel like you could use `all` here (or better just remove this and use all where you previously used this)

File: unimport/refactor.py [hunk 9]
```diff
@@ -1,137 +1,111 @@
...
+                    names_to_keep.append(
+                        import_alias.with_changes(
+                            comma=cst.MaybeSentinel.DEFAULT
+                        )
+                    )
+            if len(names_to_keep) == 0:
+                return cst.RemoveFromParent()
+            else:
+                return updated_node.with_changes(names=names_to_keep)
+        return updated_node
```
Comment (isidentical) [inline unimport/refactor.py:95]: Split and simplify this function

File: unimport/refactor.py [hunk 10]
```diff
@@ -1,137 +1,111 @@
...
 
-class RefactorTool(RefactoringTool):
-    def __init__(self):
-        self._fixer = RefactorImports()
-        self._fixers = [self._fixer]
-        super().__init__(None, options={"print_function": True})
+    def leave_ImportFrom(
+        self, original_node: cst.ImportFrom, updated_node: cst.ImportFrom
+    ) -> cst.ImportFrom:
+        return self.leave_import_alike(original_node, updated_node)
```
Comment (isidentical) [inline unimport/refactor.py:105]: or better, just use this functions instead of re-directing everything to a single place. If there is common code between them just put it to a singular function.

File: unimport/refactor.py [hunk 11]
```diff
@@ -1,137 +1,111 @@
...
-        return self._fixers, []
 
-    def refactor_string(self, data, unused_imports, name="unimport"):
-        with self._fixer.clean(unused_imports):
-            source = super().refactor_string(data, name)
-        return str(source)
+def refactor_string(scanner):
+    wrapper = MetadataWrapper(cst.parse_module(scanner.source))
+    fixed_module = wrapper.visit(RemoveUnusedImportTransformer(scanner))
+    return fixed_module.code
```
Comment (isidentical) [inline unimport/refactor.py:111]: As I said, this shouldn't take up the scanner, it should take the unused imports and the source

File: unimport/scan.py [hunk 12]
```diff
@@ -25,6 +25,7 @@ def __init__(self, source=None):
         self.imports = []
         self.classes = []
         self.functions = []
+        self.source = source
```
Comment (isidentical) [inline unimport/scan.py:28]: this is not needed

File: unimport/scan.py [hunk 13]
```diff
@@ -108,6 +109,7 @@ def visit_Attribute(self, node):
         )
 
     def run_visit(self, source):
+        self.source = source
```
Comment (isidentical) [inline unimport/scan.py:112]: ditto

File: unimport/session.py [hunk 14]
```diff
@@ -1,21 +1,20 @@
 import difflib
 import fnmatch
+import pathlib
 import tokenize
 from lib2to3.pgen2.parse import ParseError
```
Comment (isidentical) [inline unimport/session.py:5]: wait, what?

File: unimport/session.py [hunk 15]
```diff
@@ -1,21 +1,20 @@
...
-from pathlib import Path
```
Comment (isidentical) [inline unimport/session.py:5]: why did you removed this

**Comment by isidentical** [comment]
These points should be addressed before PR gets merged, I didn't mean to approve, sorry.

File: setup.py (hunk 5 above)
Comment (hakancelik96) [reply setup.py:42]: > why was this deleted

Unnecessary.

File: unimport/refactor.py (hunk 8 above)
Comment (hakancelik96) [reply unimport/refactor.py:38]: I don't think I understand what you mean.

File: unimport/refactor.py (hunk 6 above)
Comment (hakancelik96) [reply unimport/refactor.py]: fix

File: unimport/refactor.py (hunk 11 above)
Comment (hakancelik96) [reply unimport/refactor.py]: fix

File: unimport/scan.py (hunk 12 above)
Comment (hakancelik96) [reply unimport/scan.py]: fix

File: unimport/scan.py (hunk 13 above)
Comment (hakancelik96) [reply unimport/scan.py]: fix

File: unimport/session.py (hunk 14 above)
Comment (hakancelik96) [reply unimport/session.py]: fix

File: unimport/session.py (hunk 15 above)
Comment (hakancelik96) [reply unimport/session.py:5]: fix

File: requirements-dev.txt (hunk 4 above)
Comment (hakancelik96) [reply requirements-dev.txt]: fix https://github.com/unimport/unimport/pull/31/commits/513fec4a6112fb30844693ce22968254970a0c19

File: unimport/refactor.py (hunk 7 above)
Comment (hakancelik96) [reply unimport/refactor.py]: fix

File: unimport/refactor.py (hunk 9 above)
Comment (hakancelik96) [reply unimport/refactor.py]: fix

**Comment by isidentical** [comment]
I'm going a head and merging this PR, it will probably cause much more problems in the near future but ¯\_(ツ)_/¯
